staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go

```diff
@@ -533,64 +677,118 @@ func (a *Authenticator) AuthenticateToken(ctx context.Context, token string) (*a
 			var groups stringOrArray
 			if err := c.unmarshalClaim(groupsClaim, &groups); err != nil {
-				return nil, false, fmt.Errorf("oidc: parse groups claim %q: %v", groupsClaim, err)
+				return nil, fmt.Errorf("oidc: parse groups claim %q: %w", groupsClaim, err)
 			}
-			info.Groups = []string(groups)
+
+			prefix := a.jwtAuthenticator.ClaimMappings.Groups.Prefix
+			if prefix != nil && *prefix != "" {
+				for i, group := range groups {
+					groups[i] = *prefix + group
+				}
+			}
+
+			return []string(groups), nil
 		}
 	}
 
-	groupsPrefix := a.jwtAuthenticator.ClaimMappings.Groups.Prefix
-	if groupsPrefix != nil && *groupsPrefix != "" {
-		for i, group := range info.Groups {
-			info.Groups[i] = *groupsPrefix + group
+	if a.celMapper.Groups == nil {
+		return nil, nil
+	}
+
+	evalResult, err := a.celMapper.Groups.EvalClaimMapping(ctx, claimsUnstructured)
+	if err != nil {
+		return nil, fmt.Errorf("oidc: error evaluating group claim expression: %w", err)
+	}
+
+	groups, err := convertCELValueToStringList(evalResult.EvalResult)
+	if err != nil {
+		return nil, fmt.Errorf("oidc: error evaluating group claim expression: %w", err)
+	}
+	return groups, nil
+}
+
+func (a *Authenticator) getUID(ctx context.Context, c claims, claimsUnstructured *unstructured.Unstructured) (string, error) {
+	uidClaim := a.jwtAuthenticator.ClaimMappings.UID.Claim
+	if len(uidClaim) > 0 {
+		var uid string
+		if err := c.unmarshalClaim(uidClaim, &uid); err != nil {
+			return "", fmt.Errorf("oidc: parse uid claim %q: %w", uidClaim, err)
 		}
+		return uid, nil
 	}
 
-	// check to ensure all required claims are present in the ID token and have matching values.
-	for _, claimValidationRule := range a.jwtAuthenticator.ClaimValidationRules {
-		claim := claimValidationRule.Claim
-		value := claimValidationRule.RequiredValue
+	if a.celMapper.UID == nil {
+		return "", nil
+	}
 
-		if !c.hasClaim(claim) {
-			return nil, false, fmt.Errorf("oidc: required claim %s not present in ID token", claim)
+	evalResult, err := a.celMapper.UID.EvalClaimMapping(ctx, claimsUnstructured)
+	if err != nil {
+		return "", fmt.Errorf("oidc: error evaluating uid claim expression: %w", err)
+	}
+	if evalResult.EvalResult.Type() != celgo.StringType {
+		return "", fmt.Errorf("oidc: error evaluating uid claim expression: %w", fmt.Errorf("uid claim expression must return a string"))
+	}
+
+	return evalResult.EvalResult.Value().(string), nil
+}
+
+func (a *Authenticator) getExtra(ctx context.Context, claimsUnstructured *unstructured.Unstructured) (map[string][]string, error) {
+	if a.celMapper.Extra == nil {
+		return nil, nil
+	}
+
+	evalResult, err := a.celMapper.Extra.EvalClaimMappings(ctx, claimsUnstructured)
+	if err != nil {
+		return nil, err
+	}
+
+	extra := make(map[string][]string, len(evalResult))
+	for _, result := range evalResult {
+		extraMapping, ok := result.ExpressionAccessor.(*authenticationcel.ExtraMappingCondition)
```

Comment:
this cast is a code smell to clean up in 1.30 before beta... we're encapsulating the key information in the accessor, carrying it along to the evalresults, and then digging it back out here.

Reply:
Agreed! Have a task in https://github.com/kubernetes/kubernetes/issues/121553 to try and use generics.